### ISSUE TITLE
False positive for PossiblyNullReference / PossiblyNullArgument after xor

### ISSUE THREAD

**Issue by artjomsimon** [vimeo/psalm#3467]
Eliminating the possibility for a null reference and returning early in that case, Psalm still warns about a possible null reference: 

```php
<?php declare(strict_types=1);

class Greeter {
    public function sayHello() : string {
    	return 'hello';
    }
}

function isGreeterDifferent(?Greeter  $greeterA, ?Greeter $greeterB) : bool {
 
    if ($greeterA === null xor $greeterB === null) {
        return true;
    }
    
    if ($greeterA === null && $greeterB === null) {
        return false;
    }

    return $greeterA->sayHello() !== $greeterB->sayHello();
}
```

Using XOR: https://psalm.dev/r/1a8967afc2
Using AND / OR: https://psalm.dev/r/a4be25d3a1

**Comment by muglug** [comment]
Psalm is happy if you remove the last (redundant) check: https://psalm.dev/r/1f7fb3e8f7

**Comment by artjomsimon** [comment]
@muglug inetresting, thanks for the observation. Indeed that second check is unnecessary.

I refactored the example slightly and produced both, a `PossiblyNullReference` and an additional `PossiblyNullArgument`: https://psalm.dev/r/550d94e4c5

* Now, trying to work around Psalm complaining by removing the redundant check for `$greeterA` removes the `PossiblyNullReference`, but not the `PossiblyNullArgument`: https://psalm.dev/r/d875cc7bfe
* Conversely, removing the redundant check for `$greeterB` removes the `PossiblyNullArgument`, but not the `PossiblyNullReference`: https://psalm.dev/r/aae9af8c9b

**Comment by muglug** [comment]
Yeah, `xor` is not currently understood properly by the algebra simplification engine.

**Comment by artjomsimon** [comment]
Just to clarify further, this is not about the literal `xor`, but rather the *concept* of it.
It's reproducible without explicitly having `xor` in the code, just using boolean OR and AND: https://psalm.dev/r/a4be25d3a1

**Comment by orklah** [comment]
While the XOR example still issues errors, the AND example is fixed

**Comment by orklah** [comment]
> Just to clarify further, this is not about the literal `xor`, but rather the _concept_ of it. It's reproducible without explicitly having `xor` in the code, just using boolean AND: https://psalm.dev/r/5e423429d0

This example is solved as well. If someone wants support for the xor operator, please open a new issue
